File: ios/Classes/native/FlutterAdropNativeAd.swift
import Foundation
import Flutter
import AdropAds


class FlutterAdropNativeAd: NSObject, AdropAd, AdropNativeAdDelegate {

    private let messenger:FlutterBinaryMessenger
    private let adropEventListenerChannel: FlutterMethodChannel?
    private let requestId: String
    /// Invoked on `onAdReceived` for backfill ads to re-bind the PlatformView's
    /// AdropNativeAdView to the new AdMob GADNativeAd. Direct ads do not need this —
    /// they reuse the same WebView host. See `AdropNativeAdViewFactory.rebind`.
    private let nativeRebindCallback: ((String) -> Void)?
    let nativeAd: AdropNativeAd

    init(
        unitId: String,
        requestId: String,
        useCustomClick: Bool,
        preferredAdChoicesPosition: Int,
        messenger: FlutterBinaryMessenger,
        nativeRebindCallback: ((String) -> Void)? = nil) {
            self.messenger = messenger
            self.requestId = requestId
            self.nativeRebindCallback = nativeRebindCallback
            let methodChannelName = AdropChannel.adropEventListenerChannel(adType: .native, id: requestId)
            self.adropEventListenerChannel = methodChannelName != nil ? FlutterMethodChannel(name: methodChannelName!, binaryMessenger: messenger) : nil
            self.nativeAd = AdropNativeAd(unitId: unitId)
            self.nativeAd.useCustomClick = useCustomClick
            self.nativeAd.preferredAdChoicesPosition = AdropAdChoicesPosition(rawValue: preferredAdChoicesPosition) ?? .topRight
        }

    func load() {
        self.nativeAd.delegate = self
        nativeAd.load()
    }

    func show() {

    }

    func onAdClicked(_ ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_CLICK_AD, arguments: metadataOf(ad))
    }

    func onAdReceived(_ ad: AdropNativeAd) {
        // Backfill: re-bind PlatformView's AdropNativeAdView so AdMob GADNativeAdView
        // gets its nativeAd rebound. Direct ads skip — the reused WebView host updates
        // its creative HTML internally.
        if ad.isBackfilled {
            nativeRebindCallback?(requestId)
        }
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_RECEIVE_AD, arguments: metadataOf(ad))
    }

    func onAdFailedToReceive(_ ad: AdropNativeAd, _ errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAILED_TO_RECEIVE, arguments: ["errorCode":AdropErrorCodeToString(code: errorCode)])
    }

    func onAdImpression(_ ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_IMPRESSION, arguments: metadataOf(ad))
    }

    func onAdVideoStart(_ ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_VIDEO_START, arguments: metadataOf(ad))
    }

    func onAdVideoEnd(_ ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_VIDEO_END, arguments: metadataOf(ad))
    }

    func dictionaryToJSONString(_ dictionary: [String: Any]) -> String? {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dictionary, options: .prettyPrinted)
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        } catch {
            return "{}"
        }
    }

    private func metadataOf(_ ad: AdropNativeAd) -> [String: Any] {
        var creative = ad.creative
        let adPlayerCallback = "window.adPlayerVisibilityCallback"
        if creative.contains(adPlayerCallback) {
            creative = creative.replacingOccurrences(of: adPlayerCallback, with: "callback(true);\(adPlayerCallback)")
        }

        return [
            "creativeId": ad.creativeId,
            "headline": ad.headline,
            "body": ad.body,
            "displayLogo": ad.profile.displayLogo,
            "displayName": ad.profile.displayName,
            "extra": dictionaryToJSONString(ad.extra) ?? "",
            "asset": ad.asset,
            "destinationURL": ad.destinationURL ?? "",
            "creative": creative,
            "creativeSizeWidth": ad.creativeSize.width,
            "creativeSizeHeight": ad.creativeSize.height,
            "txId": ad.txId,
            "campaignId": ad.campaignId,
            "isBackfilled": ad.isBackfilled,
            "callToAction": ad.callToAction,
            "browserTarget": ad.browserTargetValue.rawValue,
            "creativeType": ad.creativeType
        ]
    }
}

File: android/src/main/kotlin/io/adrop/adrop_ads/native/FlutterAdropNativeAd.kt
package io.adrop.adrop_ads.native

import android.content.Context
import io.adrop.adrop_ads.AdType
import io.adrop.adrop_ads.AdropAd
import io.adrop.adrop_ads.bridge.AdropChannel
import io.adrop.adrop_ads.bridge.AdropMethod
import io.adrop.ads.model.AdropErrorCode
import io.adrop.ads.nativeAd.AdropAdChoicesPosition
import io.adrop.ads.nativeAd.AdropNativeAd
import io.adrop.ads.nativeAd.AdropNativeAdListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

class FlutterAdropNativeAd(
    context: Context,
    unitId: String,
    requestId: String,
    useCustomClick: Boolean,
    preferredAdChoicesPosition: Int,
    messenger: BinaryMessenger,
    /**
     * Invoked on [onAdReceived] for backfill ads to re-bind the PlatformView's
     * AdropNativeAdView to the new AdMob NativeAd. Direct ads do not need this —
     * they reuse the same WebView host. See [AdropNativeAdViewFactory.rebind].
     */
    private val nativeRebindCallback: ((String) -> Unit)? = null
): AdropAd(), AdropNativeAdListener {

    val nativeAd: AdropNativeAd = AdropNativeAd(context, unitId, "")
    private val requestId: String = requestId
    private var adropEventListenerChannel: MethodChannel?

    init {
        nativeAd.listener = this
        nativeAd.useCustomClick = useCustomClick
        // The core SDK's AdropAdChoicesPosition.fromValue gets stripped by the
        // release AAR proguard (which keeps only public methods), so map here.
        nativeAd.preferredAdChoicesPosition = when (preferredAdChoicesPosition) {
            AdropAdChoicesPosition.TOP_LEFT.value -> AdropAdChoicesPosition.TOP_LEFT
            AdropAdChoicesPosition.BOTTOM_LEFT.value -> AdropAdChoicesPosition.BOTTOM_LEFT
            AdropAdChoicesPosition.BOTTOM_RIGHT.value -> AdropAdChoicesPosition.BOTTOM_RIGHT
            else -> AdropAdChoicesPosition.TOP_RIGHT
        }
        val channelName = AdropChannel.adropEventListenerChannelOf(AdType.Native, requestId)
        adropEventListenerChannel = if (channelName != null) {
            MethodChannel(messenger, channelName)
        } else {
            null
        }
    }



    override fun load() {
        nativeAd.load()
    }

    override fun destroy() {
    }

    override fun onAdClicked(ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_CLICK_AD, metadataOf(ad))
    }

    override fun onAdFailedToReceive(ad: AdropNativeAd, errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(
            AdropMethod.DID_FAIL_TO_RECEIVE_AD,
            mapOf("errorCode" to errorCode.name)
        )
    }

    override fun onAdReceived(ad: AdropNativeAd) {
        // Backfill: re-bind PlatformView's AdropNativeAdView so AdMob NativeAdView
        // gets setNativeAd called with the new AdMob NativeAd. Direct ads skip — the
        // reused WebView host updates its creative HTML internally.
        if (ad.isBackfilled) {
            nativeRebindCallback?.invoke(requestId)
        }
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_RECEIVE_AD, metadataOf(ad))
    }

    override fun onAdImpression(ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_IMPRESSION, metadataOf(ad))
    }

    override fun onAdVideoStart(ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_VIDEO_START, metadataOf(ad))
    }

    override fun onAdVideoEnd(ad: AdropNativeAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_VIDEO_END, metadataOf(ad))
    }

    private fun metadataOf(ad: AdropNativeAd): Map<String, Any> {
        var creative = ad.creative
        val adPlayerCallback = "window.adPlayerVisibilityCallback"

        if (creative?.contains(adPlayerCallback) == true) {
            creative = creative.replace(adPlayerCallback, "callback(true);$adPlayerCallback")
        }

        return mapOf(
            "creativeId" to ad.creativeId,
            "headline" to ad.headline,
            "body" to ad.body,
            "displayLogo" to ad.profile.displayLogo,
            "displayName" to ad.profile.displayName,
            "extra" to ad.extra.toString(),
            "asset" to ad.asset,
            "destinationURL" to ad.destinationURL,
            "creative" to creative,
            "creativeSizeWidth" to ad.creativeSize.width,
            "creativeSizeHeight" to ad.creativeSize.height,
            "txId" to ad.txId,
            "campaignId" to ad.campaignId,
            "isBackfilled" to ad.isBackfilled,
            "callToAction" to ad.callToAction,
            "browserTarget" to ad.browserTarget,
            "creativeType" to ad.creativeType
        )
    }
}
